shared/src/main/scala/mlscript/JSBackend.scala

```diff
@@ -425,55 +426,45 @@ class JSBackend(pgrm: Pgrm) {
-                // ```
-                // let <tempName> = <expr>;
-                // defs.<name> = <tempName>;
-                // ```
-                JSLetDecl(tempName, translatedBody) ::
-                  JSExprStmt(
-                    JSAssignExpr(
-                      JSMember(JSIdent(defResultObjName), name),
-                      JSIdent(tempName)
-                    )
-                  ) :: Nil
+              val declName = if (classNames contains originalName) {
+                val alias = topLevelScope allocate originalName
+                ctorAliasMap += originalName -> alias
+                alias
               } else {
-                // Re-assign and record the value as a new name:
-                // ```
-                // <tempName> = <expr>;
-                // defs["<name>@<number of shadow>"] = <tempName>;
-                // ```
-                JSExprStmt(JSAssignExpr(JSIdent(tempName), translatedBody)) ::
-                  JSExprStmt(
-                    JSAssignExpr(
-                      JSMember(JSIdent(defResultObjName), shadowedName),
-                      JSIdent(tempName)
-                    )
-                  ) :: Nil
+                originalName
               }
+              // We need to save a snapshot after each computation.
+              // The snapshot name will be same as the original name, or in the
+              // format of `originalName@n` if the original name is used.
+              val snapshotName = resolveShadowName(originalName)
+              (if (snapshotName === originalName) {
+                 // First time: `let <declName> = <expr>;`
+                 JSLetDecl(declName, translatedBody)
+               } else {
+                 // Not first time: `<declName> = <expr>;`
+                 JSExprStmt(JSAssignExpr(JSIdent(declName), translatedBody))
+               }) :: // Save the value: `defs.<snapshotName> = <declName>;`
+                JSExprStmt(
+                  JSAssignExpr(
+                    JSMember(JSIdent(defResultObjName), snapshotName),
+                    JSIdent(declName)
+                  )
+                ) :: Nil
+            // Ignore type declarations.
             case Def(isRecursive, name, R(body)) => Nil
-            case _: Term                         => Nil
-          })
-          // Generate something like `exprs.push(<expr>)`.
-          .concat(otherStmts.zipWithIndex.collect { case (term: Term, index) =>
-            JSExprStmt(
-              JSInvoke(
-                JSMember(JSIdent(exprResultObjName), "push"),
-                translateTerm(term)(topLevelScope) :: Nil
-              )
-            )
+            // `exprs.push(<expr>)`.
+            case term: Term =>
+              JSExprStmt(
+                JSInvoke(
+                  JSMember(JSIdent(exprResultObjName), "push"),
+                  translateTerm(term)(topLevelScope) :: Nil
+                )
+              ) :: Nil
```

Comment:
Here I fixed a misbehaviour. Declarations and expressions should generated in order. I used to generate declarations first, and then generate expressions.